### PR TITLE
Development

### PR DESCRIPTION
**Pull request by aledj2**
This change is [Reviewable](https://reviewable.io/reviews/moka-guys/agilent_array_design_files/1)

### CODE REVIEW

File: turn_probes_off.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        # set output files
+        self.filtered_array_design="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\dev\\CGH 4.3_085030_D_F_20170809_Guys19_FILTERED.xml"
+        self.filtered_array_design_bedfile="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\dev\\CGH 4.3_085030_D_BED_20170809_FILTERED.bed"
+        
+        # add to the array design name
+        self.name_to_modify_design_name="_filtered"
+        
+    def get_probe_list(self,xml_input_file):
+        '''this function takes a single array design and returns a list of probe names'''
+        # empty list to hold all probes in this design
```
Comment (woook) [inline turn_probes_off.py]: Clarify which design

File: turn_probes_off.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        probe_list=[]
+   
+        #load the xml file into python object
```
Comment (woook) [inline turn_probes_off.py]: clarify which xml

File: turn_probes_off.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        for probe in root.iter('reporter'):
+            # add the probe name to the list
+            probe_list.append(probe.attrib['name'])
+        #return list
+        return probe_list
+    
+    def rewrite_control_file(self, oldprobelist, file_to_filter):
+        '''this function takes the lists of probes that are to be 'turned off from the new array design and rewrites the xml file'''
+        
+        #load the xml file into python object
```
Comment (woook) [inline turn_probes_off.py]: which xml?

File: turn_probes_off.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        for probe in root.iter('reporter'):
+            #if the probe name is not in the old array design
+            if probe.attrib['name'] not in oldprobelist:
+                # add a flag to ignore the probe
+                probe.attrib['control_type']="ignore"       
+                
+        #write the new modified design to file 
+        tree.write(self.filtered_array_design)
+        
+        #### Also need to write in some non-xml lines (comments and doc type)
```
Comment (woook) [inline turn_probes_off.py]: This section does not seem very robust and a bit of an after thought. Can it be it's own class?
I don't like top_50_lines but if you have to do it this way, then can you at least test that `<project ` occurs in the top 50 lines?
Alternatively, can't you just parse the top of the XML until you hit `<project `, then add that to the start of the filtered XML?

File: turn_probes_off.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            # if we have already reached the xml pass over
+            else:
+                pass
+        
+        # re-write the list with the non-xml added in to file.
+        with open(self.filtered_array_design,'w') as filtered_file_with_header:
+            filtered_file_with_header.writelines(filtered_no_header_list)
+    
+    def create_new_bedfile(self, old_probes):
+        '''This function takes the list of probes on the old array design as an input. It then opens the array_design bedfile in a list, loops through it assessing if the probe name is on the old array design. if it is the line is written to a new bedfile.'''
```
Comment (woook) [inline turn_probes_off.py]: See above about docstrings

File: turn_probes_off.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
+import xml.etree.ElementTree as ET
+from bokeh.server.start import stop
+
+class edit_design_file():
```
Comment (woook) [inline turn_probes_off.py]: Add docstring but see comments below.

File: turn_probes_off.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                probename=line.split('\t')[3].rstrip()
+                # if the probe name is on the old array write to the new bedfile
+                if probename in set(old_probes):
+                    filtered_bed.write(line)
+            
+    
+               
+        
+           
+    def what_to_do(self):
```
Comment (woook) [inline turn_probes_off.py]: Can we have a better name?

File: turn_probes_off.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        '''This function uses the above modules to create the new array file'''
+        # get lists of probes from new and old array designs
+        old_probelist=self.get_probe_list(self.old_array_design) 
+        #new_probelist=self.get_probe_list(self.new_array_design)
```
Comment (woook) [inline turn_probes_off.py]: Delete this line or add a comment to explain why it's commented out.

File: turn_probes_off.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        
+        # call function to write a design file where new probes have been filtered
+        self.rewrite_control_file(old_probelist, self.new_array_design)
+        
+        ########################################################################
```
Comment (woook) [inline turn_probes_off.py]: Delete or add a comment to explain why it's commented out.

File: turn_probes_off.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        #         in_both_count+=1
+        #     else:
+        #         new_probe+=1
+        # 
+        ########################################################################
+        
+        #create the bed file 
+        self.create_new_bedfile(old_probelist)
+        
+        ########################################################################
```
Comment (woook) [inline turn_probes_off.py]: Delete or add a comment to explain why it's commented out.

File: README.md [hunk 11]
```diff
@@ -1 +1,6 @@
-# agilent_array_design_files
\ No newline at end of file
+# agilent_array_design_files
```
Comment (woook) [inline README.md]: Either in the readme, or the docstring for class edit_design_file, please introduce your naming conventions and use them consistently throughout. Otherwise, is very hard to follow what is being done to what.
e.g.

Old design - define
New design - define
Filtered design - define

(don't write do x to this design from that design to output a design)

and then use this terminology for the BED files as well so that it's consistent and the person reading your code is helped to understand.

i.e.
Old BED - define
New BED - define
Filtered BED - define

**Comment by aledj2** [comment]
changes made

File: turn_probes_off.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    This script takes all the probes from the accredited old design and uses this to turn any probes on the new design into control probes, in effect turning the probes off.
+    The output is a "filtered" version of the "new" array design. 
+    
+    The script also creates a modified bedfile for the "new" design, removing any probes from the "new" design bed file which don't appear on the "old" array design 
+    """
+    
+    def __init__(self):
+        """Specify the input and output files"""
+        # input xml file for "old" design
+        self.OldArrayDesignXML="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\array_design_028469\\028469_Guys19_110601.xml"
```
Comment (woook) [inline turn_probes_off.py]: Capitals issue #2

File: README.md [hunk 13]
```diff
@@ -1 +1,13 @@
...
+This repository contains scripts used to search and manipulate agilent array design files
+
+## turn_probes_off.py
+An probes on an array design are not ISO accredited until the array has been accredited, eg through extension to scope. Therefore, after changing the array design, to maintain an ISO accredited service any unaccredited probes (probes not on the 'old' accredited array) cannot be used.
```
Comment (woook) [inline README.md]: This is unclear - #3

File: turn_probes_off.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        # input xml file for "new" design
+        self.new_array_design_xml="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\array_design_084720\\CGH 4.2_084720_D_F_20170309.xml"
+        
+        # input bedfile for "new" design
+        self.new_array_design_bedfile="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\array_design_085030\\CGH 4.3_085030_D_BED_20170809.bed"
+        
+        # output xml file for "filtered" design
+        self.filtered_array_design_xml="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\filtered_new_design\\CGH 4.2_084720_D_F_20170309_FILTERED.xml"
+        # output bedfile for "filtered" design
+        self.filtered_array_design_bedfile="S:\\Genetics_Data2\\Array\\Audits and Projects\\160421 ArrayCGH v4.0 Design\\modified_design_file\\filtered_new_design\\ignore_me.bed"
```
Comment (woook) [inline turn_probes_off.py]: #5

**Comment by woook** [comment]
Reviewed 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/moka-guys/agilent_array_design_files/1)*

**Comment by woook** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/moka-guys/agilent_array_design_files/1)*

**Comment by woook** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/moka-guys/agilent_array_design_files/1#-:-Ky_h8kTlmQdIdUElqx7:bnfp4nl)*
